### PR TITLE
Delay formatter setup until reporter needs to notify something

### LINKED ISSUE #2224: Adding report listener with documentation formatter results in duplicate summary
I've added a custom listener using:

```
RSpec.configure do |config|
  config.reporter.register_listener(Reporter.new, :start, :example_started, :start_dump)
end
```

It works great and when I leave the output formatter as default I see a dot per passing example and summary as expected.

However, if I change my `.rspec` config file to include `--format documentation` then I see a dot and the spec description per example and also I see duplicate summaries when the suite is complete.

My limited understanding is that when you register a listener object, it is notified of the events that you subscribe to but it isn't classed as a formatter per se. Is this expected behaviour and if so, is there a way to use the documentation formatter and a custom listener and only get one set of output?

I looked around the rspec-core codebase but didn't spot any obvious reason for this but I'm not very familiar with it. The version of rspec-core that I'm using is: 3.4.4. If it helps, here is the project and commit that contains what I'm up to: https://github.com/andypike/rectify/pull/8/commits/e61b6b10d05686782b03c8625fbb2404bc1db06a (it collects database stats as spec run).

Thanks in advance for any help and also thank you for all the work you folks do on RSpec, it's greatly appreciated ❤️ ❤️ ❤️ ❤️ ❤️

### PR DESCRIPTION
**Pull request by JonRowe**
By delaying formatter setup until the reporter needs to tell someone something we allow people to access reporter public apis without triggering doubling up on formatters.

Fixes #2224

### CODE REVIEW

**Comment by xaviershay** [comment]
@JonRowe what do you want to do with this? Seems legit?

**Comment by xaviershay** [comment]
(I'm happy to take it over and rebase etc...)

**Comment by JonRowe** [comment]
@xaviershay this should be ready for re-review

File: Gemfile [hunk 1]
```diff
@@ -30,6 +30,10 @@ if RUBY_VERSION < '2.0.0' || RUBY_ENGINE == 'java'
   gem 'json', '< 2.0.0'
 end
 
+if RUBY_VERSION < '2.0.0' && !!(RbConfig::CONFIG['host_os'] =~ /cygwin|mswin|mingw|bccwin|wince|emx/)
+  gem 'ffi', '< 1.9.15' # allow ffi to be installed on older rubies on windows
+end
```
Comment (mrageh) [inline Gemfile]: @JonRowe Why is this commit relevant to this PR?

Comment (JonRowe) [inline Gemfile]: Its not but its a requirement of getting the build to pass. It was done here first as a test and has since been merged upstream; but I didn't think it was worth rebasing just to remove it from the github UI.

File: spec/rspec/core/configuration_spec.rb [hunk 2]
```diff
@@ -1459,6 +1459,48 @@ def metadata_hash(*args)
       end
     end
 
+    describe '#reporter' do
+      before do
+        config.output_stream = StringIO.new
+        config.deprecation_stream = StringIO.new
+      end
+
+      it 'doesnt immediately trigger formatter setup' do
```
Comment (xaviershay) [inline spec/rspec/core/configuration_spec.rb]: nit: "does not" so the missing apostrophe doesn't trigger me :P

File: spec/rspec/core/configuration_spec.rb [hunk 3]
```diff
@@ -1459,6 +1459,48 @@ def metadata_hash(*args)
...
+      it 'allows registering listeners without doubling up formatters' do
+        config.reporter.register_listener double(:message => nil), :message
+
+        expect {
+          config.formatter = :documentation
+        }.to change { config.formatters.size }.from(0).to(1)
+
+        expect {
+          config.reporter.notify :message, double(:message => 'Triggers formatter setup')
+        }.to change { config.formatters.size }.from(1).to(2)
```
Comment (xaviershay) [inline spec/rspec/core/configuration_spec.rb]: The implementation of this test is weird to me. Why do you end up with 2 formatters?

File: spec/rspec/core/configuration_spec.rb [hunk 4]
```diff
@@ -1459,6 +1459,48 @@ def metadata_hash(*args)
...
+      end
+
+      it 'still configures a default formatter when none specified' do
+        config.reporter.register_listener double(:message => nil), :message
+        expect {
+          config.reporter.notify :message, double(:message => 'Triggers formatter setup')
+        }.to change { config.formatters.size }.from(0).to(2)
```
Comment (xaviershay) [inline spec/rspec/core/configuration_spec.rb]: I guess this kind of answers my question above, but kind of feels brittle that we assume there are two default formatters?

File: spec/rspec/core/formatters/base_text_formatter_spec.rb [hunk 5]
```diff
@@ -35,6 +35,11 @@
       expect(formatter_output.string).to match("1 example, 1 failure, 1 pending")
     end
 
+    it "with 1s outputs singular (only pending)" do
```
Comment (xaviershay) [inline spec/rspec/core/formatters/base_text_formatter_spec.rb]: this one is unrelated right? (Don't mind including it, just making sure I understand)

File: Changelog.md [hunk 6]
```diff
@@ -3,6 +3,8 @@
 Bug Fixes:
 
 * Include example id in the JSON formatter output. (#2369, Xavier Shay)
+* Delay formatter loading until the last minute to allow accessing the reporter
```
Comment (xaviershay) [inline Changelog.md]: LGTM overall, couple of comments but am ok with this being merged.

File: lib/rspec/core/reporter.rb [hunk 7]
```diff
@@ -53,6 +55,13 @@ def register_listener(listener, *notifications)
     end
 
     # @private
+    def prepare_default(loader, output_stream, deprecation_stream)
+      @setup_default = lambda do
+        loader.setup_default output_stream, deprecation_stream
```
Comment (xaviershay) [inline lib/rspec/core/reporter.rb]: tbh this all feels a bit janky and complicated but I don't have any better ideas :(

File: spec/rspec/core/configuration_spec.rb (hunk 3 above)
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb]: Just Deprecation to Deprecation and Documentation.

File: spec/rspec/core/configuration_spec.rb (hunk 4 above)
Comment (JonRowe) [inline spec/rspec/core/configuration_spec.rb]: Theres always a deprecation formatter too

File: spec/rspec/core/formatters/base_text_formatter_spec.rb (hunk 5 above)
Comment (JonRowe) [inline spec/rspec/core/formatters/base_text_formatter_spec.rb]: Yeah, it was for coverage reasons

**Comment by JonRowe** [comment]
I corrected the typo and added some comments explaining the formatter counts for future reference.
